nautobot_circuit_maintenance/views.py

```diff
@@ -299,3 +314,76 @@ def get(self, request, *args, **kwargs):
                 "source_type": source.__class__.__name__,
             },
         )
+
+
+def google_authorize(request, slug):
+    """View to start the Google OAuth authorization flow."""
+    # Create flow instance to manage the OAuth 2.0 Authorization Grant Flow steps.
+    notification_source = models.NotificationSource.objects.get(slug=slug)
+    source = Source.init(name=notification_source.name)
+    request.session["CLIENT_SECRETS_FILE"] = source.credentials_file
+    request.session["SCOPES"] = source.SCOPES
+    request.session["SOURCE_SLUG"] = slug
+
+    # Create flow instance to manage the OAuth 2.0 Authorization Grant Flow steps.
+    flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+        request.session["CLIENT_SECRETS_FILE"], scopes=request.session["SCOPES"]
+    )
+    # The URI created here must exactly match one of the authorized redirect URIs
+    # for the OAuth 2.0 client, which you configured in the API Console. If this
+    # value doesn't match an authorized URI, you will get a 'redirect_uri_mismatch'
+    # error.
+
+    flow.redirect_uri = (
+        request.scheme
+        + "://"
+        + request.get_host()
+        + reverse("plugins:nautobot_circuit_maintenance:google_oauth2callback")
+    )
+    authorization_url, state = flow.authorization_url(
+        # Enable offline access so that you can refresh an access token without
+        # re-prompting the user for permission. Recommended for web server apps.
+        access_type="offline",
+        # Enable incremental authorization. Recommended as a best practice.
+        include_granted_scopes="true",
+        # Asking for consent everytime is providing a refresh token that if not used only happens
+        # the first time.
+        prompt="consent",
+    )
+    # Store the state so the callback can verify the auth server response.
+    request.session["state"] = state
+
+    return redirect(authorization_url)
+
+
+def google_oauth2callback(request):
+    """View to receive the callback from Google OAuth authorization flow."""
+    # Specify the state when creating the flow in the callback so that it can
+    # verified in the authorization server response.
+    state = request.session["state"]
+    flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+        request.session["CLIENT_SECRETS_FILE"], scopes=request.session["SCOPES"], state=state
+    )
+
+    flow.redirect_uri = (
+        request.scheme
+        + "://"
+        + request.get_host()
+        + reverse("plugins:nautobot_circuit_maintenance:google_oauth2callback")
+    )
+    # Use the authorization server's response to fetch the OAuth 2.0 tokens.
+    authorization_response = request.build_absolute_uri()
+
+    flow.fetch_token(authorization_response=authorization_response)
+
+    # Store credentials in the session.
+    credentials = flow.credentials
+    notification_source = models.NotificationSource.objects.get(slug=request.session["SOURCE_SLUG"])
```

Comment:
Do we need to gracefully handle a possible `NotificationSource.DoesNotExist` exception?

Reply:
I don't see it as a possible with the current workflow, only by some unexpected error, but as commented before, adding graceful error handling is always good.

Reply:
As above - a browser or Postman could hit this view endpoint, not just Google servers, I think?